icu4c/source/common/uresbund.cpp

```diff
@@ -1910,10 +1910,74 @@ static Resource getTableItemByKeyPath(const ResourceData *pResData, Resource tab
   return resource;
 }
 
-U_CAPI UResourceBundle* U_EXPORT2 
-ures_getByKeyWithFallback(const UResourceBundle *resB, 
-                          const char* inKey, 
-                          UResourceBundle *fillIn, 
+static void createPath(CharString& path,
+                       const char* resPath,
+                       int32_t     resPathLen,
+                       const char* origResPath,
+                       int32_t     origResPathLen,
+                       const char* inKey,
+                       UErrorCode* status) {
+    // This is a utility function used by ures_getByKeyWithFallback() below.  This function builds a path from
+    // resPath and inKey, returning the result in `path`.  Originally, this function just cleared `path` and
+    // appended resPath and inKey to it, but that caused problems for horizontal inheritance.
+    //
+    // In normal cases, resPath is the same as origResPath, but if ures_getByKeyWithFallback() has followed an
+    // alias, resPath may be different from origResPath.  Not only may the existing path elements be different,
+    // but resPath may also have MORE path elements than origResPath did.  If it does, those additional path
+    // elements SUPERSEDE the corresponding elements of inKey.  So this code counts the number of elements in
+    // resPath and origResPath and, for each path element in resPath that doesn't have a counterpart in origResPath,
+    // deletes a path element from the beginning of inKey.  The remainder of inKey is then appended to
+    // resPath to form the result.  (We're not using uprv_strchr() here because resPath and origResPath may
+    // not be zero-terminated.)
```

Comment:
This... sounds plausible, and I know you are fixing various bugs, so good.
However, it also sounds like we are repairing something that got messed up earlier, and it should be possible to properly track and adjust the actual keyPath as we go.
Maybe I am wrong, and the interplay between ures_getByKeyWithFallback() and createPath() is just right, but it feels dangerously convoluted. I should take time to step through the alias-following code in a debugger and see for myself...

Anyway, let's keep this as is for now. Quite possible that you have figured it all out and I am just not understanding it yet.

Reply:
Again, it's been a while since I originally figured this out, but what I remember is that you'd get paths of different lengths in cases where the alias was at a different level than the thing you were actually looking for.  That is, if you were looking for `calendar/coptic` and it aliased over to `calendar/generic`, that worked fine, but if you were looking for `calendar/coptic/availableFormats` and followed the alias from `calendar/coptic` to `calendar/generic`, then the `resPath` you got back would be something like `generic/availableFormats` and if you just stuck things together naively, you'd end up with `calendar/generic/coptic/availableFormats`, which is of course wrong.  Or something like that-- I'm sure I have the details wrong.

In other words, this logic can probably be improved, but I think the need to match up the segments is inherent in what we're doing and not really a symptom of a bug.  But it's been a while, so maybe I'm wrong about this.